File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/bruteforce/Problem14501.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class Problem14501 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        final int N = Integer.parseInt(br.readLine());
        final int[] times = new int[N];
        final int[] money = new int[N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            times[i] = Integer.parseInt(st.nextToken());
            money[i] = Integer.parseInt(st.nextToken());
        }

        final int[] dp = new int[N + 1];
        for (int i = 0; i < N; i++) {
            if (i + times[i] <= N) {
                dp[i + times[i]] = Math.max(dp[i + times[i]], dp[i] + money[i]);
            }

            dp[i + 1] = Math.max(dp[i + 1], dp[i]);
        }

        System.out.println(dp[N]);
    }
}

File: algorithm/kotlin-algorithm/baekjoon/src/main/kotlin/algorithm/baekjoon/Problem14501.kt
package algorithm.baekjoon

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

/**
 * <p>
 * 이름: 퇴사
 * 난이도: 실버-3
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/14501">퇴사 (실버-3)</a>
 **/
class Problem14501 {
    fun solution(schedules: Array<Pair<Int, Int>>): Int {
        val n = schedules.size
        val dp = IntArray(n + 1)

        for (i in schedules.indices) {
            val (t, p) = schedules[i]

            if (i + t <= n) {
                dp[i + t] = dp[i + t].coerceAtLeast(dp[i] + p)
            }
            dp[i + 1] = dp[i + 1].coerceAtLeast(dp[i])
        }
        return dp[n]
    }
}

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()
    val schedules = Array(n) {
        val st = StringTokenizer(readLine())
        st.nextToken().toInt() to st.nextToken().toInt()
    }

    println(Problem14501().solution(schedules))
}
